### PR TITLE
Refactor FileAPI implementation

### PR DESCRIPTION
**Pull request by izgzhen**
Most are simple refactoring, cleanups and improvements, but still involving two slightly notable changes:
- In `filemanager`, now we read the file content based on requested `RelativePos` by `seek` and `read_exact` (rather than `read_to_end` then do slicing). This strategy might be again adjusted in future performance tuning but certainly better than nothing.
- Also, I cached more file meta-info in both sides and left a block of comment on `filemanager`'s file reading mentioning the snapshot-state problem (not solved now though).

r? @Manishearth 

Part of #11131

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12406)

### CODE REVIEW

**Comment by Manishearth** [comment]
Overall LGTM -- could you add doc comments on all the structs and sprinkled throughout the codebase? The fact that BlobBuf can be a seek'd slice too is a bit non-obvious.

**Comment by izgzhen** [comment]
Good, I will have a double-check

**Comment by izgzhen** [comment]
Done

File: components/net/filemanager_thread.rs [hunk 1]
```diff
@@ -214,55 +233,22 @@ impl<UI: 'static + UIProvider> FileManager<UI> {
...
+                spawn_named("load blob".to_owned(), move || load_blob(load_data, consumer, classifier, blob_buf));
             }
+            Err(e) => send_error(load_data.url.clone(), NetworkError::Internal(format!("{:?}", e)), consumer),
         }
     }
 }
 
+/// File manager's data store. It maintains a thread-safe mapping
+/// from FileID to FileStoreEntry which might have different backend implementation.
+/// Access to the content is sealed as methods of this struct.
```
Comment (Manishearth) [inline components/net/filemanager_thread.rs]: s/sealed/encapsulated

**Comment by izgzhen** [comment]
Fixed

**Comment by izgzhen** [comment]
@Manishearth Sorry, I made a mistake in making `TFPProvider` private. I will r=you when this get fixed

**Comment by Manishearth** [comment]
Yep, just noticed that :smile:

@bors-servo r- delegate+

**Pull request merged**